### PR TITLE
Keychain property wrapper

### PR DESCRIPTION
**Pull request by mikolasstuchlik**
This Pull Request represents first step in my effort to improve property wrapper. Following steps will ... follow:
- Create documentation for new wrapper
- Expand unit tests
- Create example application for keychain wrapper
- **Implement coding**

What is in this Pull Request? Working example of property wrapper, that is only working for type `Data` as it's generic parameter. This limitation is due to lack of coding support. This feature proved to be harder than originally anticipated.

Coding's future support will be implemented in class  `KeychainItemPropertyWrapper`. API of the property wrapper should not change however. This Pull Request is intended to create stable and accepted API.

**General ideas behind this implementation.**

What features are expected from this property wrapper:
- Ability to access Internet Password and Generic Password keychain items
- Ability to store any type (preferably String in Internet Password) **Not implemented yet**
- Ability to access and modify those item's respective attributes
- Ability to secure Generic Password item using password, biometric and/or any other available method
- API accessible with property wrapper
- Store is accessed explicitly by programmer or when determined by property wrapper

Since password items share significant amount of same code and attributes, oop was chosen as the main decomposition method. 

Base class `KeychainItem ` is responsible for
- Defining internal API for overrides
- Composing and executing queries
- ~For better decomposition, attributes are stored in multiple structs~

Derivate class `KeychainItemPropertyWrapper` is responsible for
- Coding **Not implemented yet**
- Executing loads and stores at the right time depending on setting using `KeychainDataRefreshPolicy `

final classes `GenericPassword` and `InternetPassword` are responsible for
- Storing their exclusive attributes
- Loading them from store
- Composing them into store queries

**Reviewer notice**
This Property Wrapper does not present all features nor attributes of respective item classes, this is not an aim. Aim of this property wrapper is to have as easy as possible API which includes feature to inspect and modify selected attributes. Reviewer is kindly asked to aim most of his/her effort on public API. Any ideas to rework public API in most brutal manner which will result in better programmer experience is welcomed.
At the same time, this property wrapper is designed in a way, that will make possible to implement other attributes, platforms or queries, that are not possible today. Different platforms may require different access patterns. Pattern chosen for this property wrapper is designed with iOS in mind.

### CODE REVIEW

**Comment by mkj-is** [comment]
Closes #7.

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainCommons.swift [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        if let data = searchResult[kSecValueData as String] as? Data {
+            itemData = data
+        } else {
+            print("FTPropertyWrappers KeychainItem insertQuery: warning: update request without data")
+        }
+    }
+
+    private var insertQuery: [String: Any] {
+        var query = itemClassIdentity.merging(itemAttributes) { lhs, rhs in
+            print("FTPropertyWrappers KeychainItem insertQuery: notice: collision found at instance \(self) between \(lhs) and \(rhs)")
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainCommons.swift]: Is this programmer error or does it event matter if you override the key? I would either use assert or no logging.

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainCommons.swift [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        query[kSecMatchLimit as String] = kSecMatchLimitOne
+
+        if query[kSecReturnAttributes as String] != nil {
+            print("FTPropertyWrappers KeychainItem fetchQuery: warning: changing of kSecReturnAttributes is not allowed!")
+        }
+        query[kSecReturnAttributes as String] = true
+
+        if query[kSecReturnData as String] != nil {
+            print("FTPropertyWrappers KeychainItem fetchQuery: warning: changing of kSecReturnData is not allowed!")
+        }
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainCommons.swift]: Are those programmer errors? If yes, I would assert or add preconditions.

File: Sources/FTPropertyWrappers/InternetPassword.swift [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    
+    override public var itemClassIdentity: [String : Any] {
+        return [
+            kSecClass as String: kSecClassInternetPassword,
+            kSecAttrServer as String: serverIdentifier
+        ]
+    }
+
+    override public var wrappedValue: T? {
+        get { return super.wrappedValue }
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/InternetPassword.swift]: You can use implicit return in all the code. After all property wrappers are only available in Swift 5.1+.

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainPropertyWrapper.swift [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    override open var itemData: Data {
+        get {
+            guard let cached = cachedValue as? Data else { fatalError("Not a Data") }
+            return cached
+        }
+        set {
+            guard let newVal = newValue as? T else { fatalError("Not a Data") }
+            cachedValue = newVal
+        }
+    }
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainPropertyWrapper.swift]: Some TO-DO or note here? This code deserves it.

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainPropertyWrapper.swift [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        do {
+            try executeInsertQuery()
+            synced = true
+        } catch (KeychainError.osSecureDuplicitItem){
+            try executeUpdateQuery()
+            synced = true
+        }
+
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainPropertyWrapper.swift]: I am not sure abbout all those `synced = true` lines. I believe the control flow could be improved.

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainPropertyWrapper.swift [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        switch currentStatus {
+        case .success:
+            synced = true
+        case .failure(KeychainError.osSecureNoSuchItem):
+            synced = true
+            cachedValue = nil
+        case .failure(let error):
+            throw error
+        }
+
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainPropertyWrapper.swift]: Yoou can call this in a much more simple way:

```swift
let currentStatus = Result(catching: executeFetchQuery)
```

But still, there is no reason to use Result. Simple do-catch block would suffice.

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainTypes.swift [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+protocol ConfiguringElement {
+    func insertParameters(into query: inout [String : Any])
+    func readParameters(from response: [String : Any])
+
+    var constraints: [Constraint] { get }
+    var key: String { get }
+}
+
+public enum Constraint {
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainTypes.swift]: Ideally find a more specific name.

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+
+    private func encode(_ value: UInt32) throws -> Data {
+        var data = UInt64(value)
+        return Data(bytes: &data, count: MemoryLayout<UInt64>.size)
+    }
+
+    private func encode(_ value: UInt64) throws -> Data {
+        var data = value
+        return Data(bytes: &data, count: MemoryLayout<UInt64>.size)
+    }
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift]: I would remove throws keyword from all not throwing methods. The functions don't need to have the same signature.

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+        case is UInt32.Type:
+            return (try decode(from: data) as UInt32) as! T
+        case is UInt64.Type:
+            return (try decode(from: data) as UInt64) as! T
+        case is Float.Type:
+            fatalError("Encoding root type Float not supported!")
+        case is Float80.Type:
+            fatalError("Encoding root type Float80 not supported!")
+        case is Double.Type:
+            fatalError("Encoding root type Double not supported!")
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift]: You are mixing fatal errors with exceptions. I would unify the error handling system.

```suggestion
            throws DecodingError.dataCorrupted(
                DecodingError.Context(codingPath: [], debugDescription: "Encoding root type Double not supported!", underlyingError: nil)
            )
```

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift [hunk 10]
```diff
@@ -0,0 +1,371 @@
+import Foundation
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift]: Can we separate encoding and decoding into separate types? To respect single-responsibility principle.

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+        return try encode(value.absoluteString)
+    }
+
+    private func encode(_ value: Data) throws -> Data {
+        return value
+    }
+
+    // MARK: - Decoders overloaded
+
+    private func decode(from data: Data) throws -> Int {
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift]: What if we used the same signature for decoding as a `SingleValuedecodingContainer`? There is a reason behind it, because you are helping the type system. And if we want to support standard-library coding we will have less work to do.

https://developer.apple.com/documentation/swift/singlevaluedecodingcontainer

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+        fatalError("Encoding root type Float not supported!")
+    }
+
+    private func decode(from data: Data) throws -> Float80 {
+        fatalError("Encoding root type Float80 not supported!")
+    }
+
+    private func decode(from data: Data) throws -> Double {
+        fatalError("Encoding root type Double not supported!")
+    }
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift]: These functions are never called.

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+
+    private func decode(from data: Data) throws -> Bool {
+        let value: Int = try decode(from: data)
+        switch value {
+        case 0:
+            return false
+        case 1:
+            return true
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift]: What if we used C-style booleans? Zero is true and anything else is false?

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainCommons.swift [hunk 14]
```diff
@@ -0,0 +1,151 @@
+import Foundation
+
+open class SingleValueKeychainItem {
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainCommons.swift]: Could you please rename this file to `SingleValueKeychainItem.swift`. It only contains this one type.

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainPropertyWrapper.swift [hunk 15]
```diff
@@ -0,0 +1,93 @@
+import Foundation
+
+open class KeychainItemPropertyWrapper<T: Codable>: SingleValueKeychainItem {
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainPropertyWrapper.swift]: The name of the file does not match with the name of the file. Could you please rename it?

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainTypes.swift [hunk 16]
```diff
@@ -0,0 +1,123 @@
+//
+//  File.swift
+//  
+//
+//  Created by Mikoláš Stuchlík on 15/02/2020.
+//
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainTypes.swift]: Could you please remove this forgotten header comment?

File: Tests/FTPropertyWrappersTests/KeychainCoderTest.swift [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+        let wrapped: T
+        init(wrapped: T) {
+            self.wrapped = wrapped
+        }
+        func testCoding() throws {
+            let coder = KeychainCoding()
+            let data = try coder.encode(wrapped)
+            XCTAssertEqual(try coder.decode(T.self, from: data), wrapped)
+        }
+    }
```
Comment (mkj-is) [inline Tests/FTPropertyWrappersTests/KeychainCoderTest.swift]: There is no need for this wrapper class in tests. I think the less abstraction in test the better. I would use only the helper function.

File: Tests/FTPropertyWrappersTests/KeychainCoderTest.swift [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+            try assertCoding(input: UInt16.max)
+            try assertCoding(input: UInt16.min)
+
+            try assertCoding(input: UInt32.max)
+            try assertCoding(input: UInt32.min)
+
+            try assertCoding(input: UInt64.max)
+            try assertCoding(input: UInt64.min)
+        } catch {
+            XCTFail(error.localizedDescription)
```
Comment (mkj-is) [inline Tests/FTPropertyWrappersTests/KeychainCoderTest.swift]: Let's asset that no coding should throw errors.

```suggestion
            XCTAssertNoThrow(throws error)
```

File: Tests/FTPropertyWrappersTests/KeychainTests.swift [hunk 19]
```diff
@@ -1,87 +1,102 @@
...
-        XCTAssertEqual(tester2.number, 15)
-        tester2.number = 30
-        XCTAssertEqual(tester.number, 30)
-        XCTAssertEqual(tester2.number, 30)
+    /*
+     *  Notice: Accesible and AccessControl are not being tested in unit tests.
+     *  Those features are implemented in example apps.
+     */
+
+    func testAttribute_kSecAttrDescription() {
```
Comment (mkj-is) [inline Tests/FTPropertyWrappersTests/KeychainTests.swift]: This looks very incomplete. Can we add at least some TO-DO or add this in separate PR?

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift (hunk 13 above)
Comment (mikolasstuchlik) [reply Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift]: Treating booleans as a number is considered a bad thing even amongst C developers (I've met). Boolean is a logical value, not a number. We encode "true" as 1, not as "any positive number". Since we know, that bool is encoded by this encoder every single time as 0 or 1 (or any other pair of constants) it makes sense to indicated other values as errors - which they are.

File: Tests/FTPropertyWrappersTests/KeychainCoderTest.swift (hunk 17 above)
Comment (mikolasstuchlik) [reply Tests/FTPropertyWrappersTests/KeychainCoderTest.swift]: Having this type is a result of development proces, during which I found out, that simply having overloaded function is not a solution when called from within a generic type instance. Getting rid of this wrapper may result in coding working in tests but failing in Keychain (as it did during development). Calling overloaded method from a non-generic scope will work. Calling overloaded method from a generic scope won't work as expected.

File: Tests/FTPropertyWrappersTests/KeychainTests.swift (hunk 19 above)
Comment (mikolasstuchlik) [reply Tests/FTPropertyWrappersTests/KeychainTests.swift]: It is :) I've continued preparing tests to avoid idle time :)

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift [hunk 20]
```diff
@@ -0,0 +1,371 @@
...
+        return try encode(value.absoluteString)
+    }
+
+    private func encode(_ value: Data) -> Data {
+        return value
+    }
+
+}
+
+struct KeychainDecoder {
```
Comment (mkj-is) [inline Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift]: What about having two separate files?

File: Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift (hunk 13 above)
Comment (mkj-is) [reply Sources/FTPropertyWrappers/Keychain - Support/KeychainCoding.swift]: That goes against thing you said earlier, that the property encoder should allow simply getting values from previous versions of apps. This opens another question I forgot to ask: Do we plan to support custom encoding/decoding?

File: Tests/FTPropertyWrappersTests/KeychainCoderTest.swift [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+@testable import FTPropertyWrappers
+
+final class KeychainCoderTest: XCTestCase {
+
+    private class WrapperClass<T: Codable & Equatable>{
+
+        let wrapped: T
+        init(wrapped: T) {
+            self.wrapped = wrapped
+        }
```
Comment (mkj-is) [inline Tests/FTPropertyWrappersTests/KeychainCoderTest.swift]: Is this needed? The initializer is not generated due to private keyword?

File: Tests/FTPropertyWrappersTests/KeychainTests.swift (hunk 19 above)
Comment (mkj-is) [reply Tests/FTPropertyWrappersTests/KeychainTests.swift]: I am not going to merge uncomplete code. What is the plan with it?
